src/environmentd/tests/server.rs

```diff
@@ -1638,3 +1639,228 @@ fn test_cancel_ws() {
     }
     handle.join().unwrap();
 }
+
+#[mz_ore::test]
+#[cfg_attr(miri, ignore)] // too slow
+fn smoketest_webhook_source() {
+    let server = util::start_server(util::Config::default()).unwrap();
+    server.enable_feature_flags(&["enable_webhook_sources"]);
+
+    let mut client = server.connect(postgres::NoTls).unwrap();
+
+    #[derive(Debug, PartialEq, Eq, Deserialize, Serialize)]
+    struct WebhookEvent {
+        ts: u128,
+        name: String,
+        attrs: Vec<String>,
+    }
+
+    // Create a webhook source.
+    client
+        .execute(
+            "CREATE SOURCE webhook_json FROM WEBHOOK BODY FORMAT JSON",
+            &[],
+        )
+        .expect("failed to create source");
+
+    let now = || {
+        std::time::SystemTime::now()
+            .duration_since(std::time::UNIX_EPOCH)
+            .expect("time went backwards")
+            .as_nanos()
+    };
+
+    // Generate some events.
+    const NUM_EVENTS: i64 = 100;
+    let events: Vec<_> = (0..NUM_EVENTS)
+        .map(|i| WebhookEvent {
+            ts: now(),
+            name: format!("event_{i}"),
+            attrs: (0..i).map(|j| format!("attr_{j}")).collect(),
+        })
+        .collect();
+
+    let http_client = Client::new();
+    let webhook_url = format!(
+        "http://{}/api/webhook/materialize/public/webhook_json",
+        server.inner.http_local_addr()
+    );
+    // Send all of our events to our webhook source.
+    for event in &events {
+        let resp = http_client
+            .post(&webhook_url)
+            .json(event)
+            .send()
+            .expect("failed to POST event");
+        assert!(resp.status().is_success());
+    }
+
+    // Wait for the events to be persisted.
+    let mut attempts = 0;
+    while attempts < 10 {
```

Comment:
Use mz_ore::retry here instead.

Reply:
Great idea, don't know why I didn't do this earlier